File: MovieousDemo/Components/Shared/MDFrameView.m
//
//  MDFrameView.m
//  MovieousDemo
//
//  Created by Chris Wang on 2019/2/27.
//  Copyright © 2019 Movieous Team. All rights reserved.
//

#import "MDFrameView.h"

@implementation MDFrameView

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        _frameWidth = 1;
    }
    return self;
}

- (void)layoutSubviews {
    self.backgroundColor = [UIColor clearColor];
    [super layoutSubviews];
    [self setNeedsDisplay];
}

- (void)drawRect:(CGRect)rect {
    // Drawing code
    CGContextRef ctx = UIGraphicsGetCurrentContext();
    [UIColor.whiteColor setStroke];
    CGContextSetLineWidth(ctx, _frameWidth);
    NSInteger halfFrameWidth = _frameWidth / 2;
    UIBezierPath *path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(_margin, _margin + halfFrameWidth)];
    [path addLineToPoint:CGPointMake(rect.size.width - _margin, _margin + halfFrameWidth)];
    [path moveToPoint:CGPointMake(rect.size.width - _margin - halfFrameWidth, _margin + halfFrameWidth)];
    [path addLineToPoint:CGPointMake(rect.size.width - _margin - halfFrameWidth, rect.size.height - _margin)];
    [path moveToPoint:CGPointMake(rect.size.width - _margin - halfFrameWidth, rect.size.height - _margin - halfFrameWidth)];
    [path addLineToPoint:CGPointMake(_margin, rect.size.height - _margin - halfFrameWidth)];
    [path moveToPoint:CGPointMake(_margin + halfFrameWidth, rect.size.height - _margin - halfFrameWidth)];
    [path addLineToPoint:CGPointMake(_margin + halfFrameWidth, _margin)];
    CGContextAddPath(ctx, path.CGPath);
    CGContextSetBlendMode(ctx, kCGBlendModeCopy);
    CGContextStrokePath(ctx);
}

- (void)setFrameWidth:(CGFloat)frameWidth {
    _frameWidth = frameWidth;
    [self setNeedsDisplay];
}

- (void)setMargin:(CGFloat)margin {
    _margin = margin;
    [self setNeedsDisplay];
}

@end

File: MovieousDemo/Common/MDFrameView.swift
//
//  MDFrameView.swift
//  MovieousDemo
//
//  Created by Chris Wang on 2019/7/7.
//  Copyright © 2019 Movieous Team. All rights reserved.
//

import UIKit

class MDFrameView: UIView {
    var frameColor: UIColor = .white {
        didSet {
            self.setNeedsDisplay()
        }
    }
    
    var margin: CGFloat = 0 {
        didSet {
            self.setNeedsDisplay()
        }
    }
    
    var frameWidth: CGFloat = 1 {
        didSet {
            self.setNeedsDisplay()
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = .clear
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        self.setNeedsDisplay()
    }
    
    override func draw(_ rect: CGRect) {
        let ctx = UIGraphicsGetCurrentContext()
        self.frameColor.setStroke()
        ctx?.setLineWidth(self.frameWidth)
        let halfFrameWidth = self.frameWidth / 2
        let path = UIBezierPath()
        path.move(to: .init(x: self.margin, y: self.margin + halfFrameWidth))
        path.addLine(to: .init(x: rect.size.width - self.margin, y: self.margin + halfFrameWidth))
        path.move(to: .init(x: rect.size.width - self.margin - halfFrameWidth, y: self.margin + halfFrameWidth))
        path.addLine(to: .init(x: rect.size.width - self.margin - halfFrameWidth, y: rect.size.height - self.margin))
        path.move(to: .init(x: rect.size.width - self.margin - halfFrameWidth, y: rect.size.height - self.margin - halfFrameWidth))
        path.addLine(to: .init(x:self.margin, y: rect.size.height - self.margin - halfFrameWidth))
        path.move(to: .init(x: self.margin + halfFrameWidth, y: rect.size.height - self.margin - halfFrameWidth))
        path.addLine(to: .init(x: self.margin + halfFrameWidth, y: self.margin))
        ctx?.addPath(path.cgPath)
        ctx?.setBlendMode(.copy)
        ctx?.strokePath()
    }
}
